src/Chainweb/CutDB.hs

```diff
@@ -516,51 +504,76 @@ synchronizeProviders logger wbh providers c = do
-            else do
-                liftIO $ logFunctionText loggr Info
+
+              -- We are done
+              --
+              then return ()
+
+              -- We are not in sync and need to resolve (could be a fork a
+              -- payload provider that is not caught up yet)
+              --
+              else do
+                pLog Info
                     $ "resolving fork on startup, from " <> brief r
                     <> " to " <> brief (_forkInfoTargetState finfo)
+
+                -- Query the trace from the fork point to the target block
+                --
                 bhdb <- liftIO $ getWebBlockHeaderDb wbh cid
                 let ppRBH = _syncStateRankedBlockHash $ _consensusStateLatest r
                 ppBlock <- liftIO (lookupRankedM bhdb (int $ _rankedHeight ppRBH) (_ranked ppRBH)) `catch` \case
                     e@(TreeDbKeyNotFound {} :: TreeDbException BlockHeaderDb) -> do
-                        liftIO $ logFunctionText loggr Warn $ "PP block is missing: " <> brief ppRBH <> ", error: " <> sshow e
+                        pLog Warn $ "PP block is missing: " <> brief ppRBH <> ", error: " <> sshow e
                         MaybeT $ return Nothing
                     _ -> empty
 
+                -- FIXME: this stream can be very long. We should limit it and
+                -- proceed iteratively if necessary.
+                -- Ideally, we check the length based on block heights before
+                -- we compute the full trace below. We also have to be careful,
+                -- that we compute the fork point only once and iterate from
+                -- there, otherwise the complexity would become quadratic.
                 (forkBlocksDescendingStream S.:> forkPoint) <- liftIO $
                         S.toList $ branchDiff_ bhdb ppBlock hdr
                 let forkBlocksAscending = reverse $ snd $ partitionHereThere forkBlocksDescendingStream
-                let newTrace =
-                        zipWith
-                            (\prent child ->
-                                ConsensusPayload (view blockPayloadHash child) Nothing <$
-                                    blockHeaderToEvaluationCtx (Parent prent))
-                            (forkPoint : forkBlocksAscending)
-                            forkBlocksAscending
-                let newForkInfo = finfo { _forkInfoTrace = newTrace }
+                let newTrace = zipWith
+                        (\prent child ->
+                            ConsensusPayload (view blockPayloadHash child) Nothing <$
+                                blockHeaderToEvaluationCtx (Parent prent))
+                        (forkPoint : forkBlocksAscending)
+                        forkBlocksAscending
+
+                let newForkInfo = finfo
+                        { _forkInfoTrace = newTrace
+                        , _forkInfoBasePayloadHash =
+                            Parent (view blockPayloadHash forkPoint)
```

Comment:
This change fixes payload provider initialization.

If the `forkInfoTrace` changes, the `forkInfoBasePayloadHash` needs to be adjusted.